### PR TITLE
Implement stateless wrapper.

### LINKED ISSUE #104: Stateless Transformation
Right now, there's some boilerplate for defining simple gradient transformations. If a user wants to implement custom weight decay, clipping, constraints, etc without any state, they still have to define a nested function with an `init` that does nothing, and handle the empty state.

I think it'd be convenient to provide a `stateless` transformation that accepts a function to apply to the updates and params. We can also do the `jax.tree_multimap` for the user by default.

```python
weight_decay = optax.stateless(lambda g, p: g + 0.1 * p)
```

If the user wants to define a function that does the tree_multimap themselves:

```python
def my_function(updates, params):
    return jax.tree_multimap(lambda g, p: ..., updates, params)

optim = optax.stateless(my_function, on_leaves=False)
```

In my view, this is a very clean way to implement simple stateless transformations. I think the `on_leaves` argument could use a better name, though. I'd be happy to implement this if it sounds reasonable.

### PR DESCRIPTION
**Pull request by n2cholas**
Closes #104. 

Not sure if this belongs in wrappers.py, please let me know if I should move it elsewhere.

### CODE REVIEW

**Review by mkunesch — CHANGES_REQUESTED** [review]
Hey! Thank you so much for this PR!

I'll do a detailed review tomorrow, but I have two quick high-level questions/comments:

- I think this code could actually be part of `base.py`. It's not really a wrapper in that it doesn't wrap a gradient transformation; instead, I think it's a really fundamental tool for creating gradient transformation so could go into `base.py`. What do you think?
- Could we replace `StatelessState` by `base.EmptyState` or is there an advantage to `stateless` having its own state?

Thanks a lot again and as I said, I'll take a detailed look tomorrow!

**Comment by n2cholas** [comment]
- I agree, I'll move it to `base.py`.
- Agree with this as well, completely forgot about `base.EmptyState`!

**Review by rosshemsley — CHANGES_REQUESTED** [review]
Thanks! I like this idea, I can imagine using this for creating quick transformations :) 

I think I would be inclined to simply this a bit, and to remove the on_leaves part, (and also, I don't htink we need to check for params being None).


We could also add an explicit example to the docs of using this as a decorator:

```python

def stateless_gradient_transformation(f) -> optax.GradientTransformation:
  def init_fn(_):
    return optax.EmptyState()

  def update_fn(updates, state, params=None):
      return f(updates, params), state

  return optax.GradientTransformation(init_fn, update_fn)


@stateless_gradient_transformation
def double_grads(updates, params):
  return jax.tree_map(lambda x: 2*x, updates)


opt = optax.chain(optax.adam(1e-4), double_grads)
state = opt.init(jax.numpy.array([]))
```

What do you think?

**Comment by rosshemsley** [comment]
(@mkunesch pointed out to me that there's already a thread on this in the issues - sorry I hadn't read that! I'll let you both figure out what to do here :) )

File: optax/_src/base.py [hunk 1]
```diff
@@ -108,3 +108,41 @@ def update_fn(updates, state, params=None):
     return jax.tree_map(jnp.zeros_like, updates), state
 
   return GradientTransformation(init_fn, update_fn)
+
+
+def stateless(
+    f: Union[
+      Callable[[Updates, Optional[Params]], Updates],
+      Callable[[chex.Array, Optional[chex.Array]], chex.Array]],
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py:116]: Agreed, will do!

File: optax/_src/base.py [hunk 2]
```diff
@@ -108,3 +108,41 @@ def update_fn(updates, state, params=None):
     return jax.tree_map(jnp.zeros_like, updates), state
 
   return GradientTransformation(init_fn, update_fn)
+
+
+def stateless(
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py:113]: I think `optax.stateless` matches the naming conventions of the other library components. None of the other GradientTransformation mechanisms actually say `gradient_transformation` (e.g. `optax.chain`, `optax.mask`). What do you think?

File: optax/_src/base.py [hunk 3]
```diff
@@ -108,3 +108,41 @@ def update_fn(updates, state, params=None):
...
+    f: Union[
+      Callable[[Updates, Optional[Params]], Updates],
+      Callable[[chex.Array, Optional[chex.Array]], chex.Array]],
+    on_leaves: bool=False
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py:117]: That's a good point. I mainly envisioned this utility being used more with inline lambda functions than standalone functions. e.g.

```py
tx = optax.chain(
   ...,
   optax.stateless(lambda g, p: ..., on_leaves=True),
   ..., 
)
```

Which would be needlessly verbose without `on_leaves`:

```py
tx = optax.chain(
   ...,
   optax.stateless(lambda g_, p_: jax.tree_map(lambda g, p: ..., g_, p_)),
   ..., 
)
```

I do like your example though. Perhaps one way to go about this is to introduce two utilities: `stateless` and `stateless_tree_map`, where the latter applies the tree_map for you (though that second name could use work 😅). What do you think?

File: optax/_src/base.py [hunk 4]
```diff
@@ -108,3 +108,41 @@ def update_fn(updates, state, params=None):
...
+) -> GradientTransformation:
+  """Convenience wrapper to create a stateless GradientTransformation.
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py:119]: Will do.

File: optax/_src/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py]: Done!

File: optax/_src/base.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py]: Updated description to remove the word "wrapper" altogether since it is not a wrapper.

File: optax/__init__.py [hunk 5]
```diff
@@ -36,6 +36,7 @@
 from optax._src.base import Params
 from optax._src.base import Schedule
 from optax._src.base import set_to_zero
+from optax._src.base import stateless
```
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/__init__.py:39]: I added it to `api.rst` in its own section. Is that ok?

**Comment by n2cholas** [comment]
Thanks @rosshemsley for the feedback! I resolved a couple of your comments and left responses to others in the code review.

File: optax/_src/base_test.py [hunk 6]
```diff
@@ -59,5 +63,44 @@ def test_set_to_zero_is_stateless(self):
...
+        return jax.tree_map(lambda g_, p_: g_ + 0.1 * p_, g, p)
+
+    opt = base.stateless(weight_decay, on_leaves=on_leaves)
+    state = opt.init(params)
+    update_fn = self.variant(opt.update)
+    new_updates, new_state = update_fn(updates, state, params)
+    expected_updates = {'a': jnp.ones((1, 2)), 'b': jnp.array([2.1])}
+    chex.assert_trees_all_close(new_updates, expected_updates)
+
+      # Check repeated application
```
Comment (mkunesch) [inline optax/_src/base_test.py:89]: Nit: Do you think it is necessary to check a repeated application? Since the transformation is stateless I think one application should suffice. While in theory testing more never hurts, it also makes this test longer (harder to see exactly which behaviour is being tested), less of a unit test that checks one behaviour, and might suggest that there is something more complicated going on in a repeated step.

Comment (mkunesch) [reply optax/_src/base_test.py:89]: If you end up removing this, I think there would be so little boilerplate in the test that I'd also split it into two separate versions for with/without the tree map to keep things simple, especially if you have different function names rather than a boolean `on_leaves` flag.

File: optax/_src/base_test.py [hunk 7]
```diff
@@ -59,5 +63,44 @@ def test_set_to_zero_is_stateless(self):
...
+
+    if on_leaves:
+      def weight_decay(g, p):
+        return g + 0.1 * p
+    else:
+      def weight_decay(g, p):
+        return jax.tree_map(lambda g_, p_: g_ + 0.1 * p_, g, p)
+
+    opt = base.stateless(weight_decay, on_leaves=on_leaves)
+    state = opt.init(params)
```
Comment (mkunesch) [inline optax/_src/base_test.py:83]: I think we could guarantee to the user that `stateless` accepts and returns an `EmptyState()`. This could be done by using `EmptyState()` directly in the update function rather than calling `init` and introducing a very simple separate test for `init` (just checking it returns EmptyState).

**Review by mkunesch — CHANGES_REQUESTED** [review]
Hi! I think this looks great - thanks a lot! I added comments on the choice of names and documentation layout and some minor thoughts on the test.
Thanks again!

File: optax/_src/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py:113]: Great, I'll stick to it!

File: optax/_src/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/_src/base.py:115]: Agreed, I like the name as well `stateless_with_tree_map`.

File: optax/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (n2cholas) [reply optax/__init__.py:39]: Done! I put them next to `EmptyState`, please let me know if you think elsewhere would be better!

File: optax/_src/base_test.py (hunk 6 above)
Comment (n2cholas) [reply optax/_src/base_test.py]: Don't think it's necessary; I removed the extra check.

File: optax/_src/base_test.py (hunk 7 above)
Comment (n2cholas) [reply optax/_src/base_test.py:77]: Agreed and done!

File: optax/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkunesch) [reply optax/__init__.py:39]: Thanks! Otherwise it looks to me like the `Optax Transforms and States` section is alphabetical so I think we should probably stick to that. WDYT?

**Review by mkunesch — COMMENTED** [review]
Thanks for making all the changes, this looks great! The only comment I have is on the positioning in the documentation but otherwise LGTM!

**Comment by n2cholas** [comment]
Agreed, fixed!

**Comment by n2cholas** [comment]
Would it be possible to rerun the tests? The failures were in `control_variates_test.py`, which are unaffected by this PR.

**Review by mkunesch — APPROVED** [review]
Thanks! I have triggered the test workflow again now that we have fixed the broken test you mentioned.

File: optax/_src/base_test.py [hunk 8]
```diff
@@ -15,8 +15,11 @@
 """Tests for base.py."""
 
 from absl.testing import absltest
+
 import chex
 import numpy as np
```
Comment (mkunesch) [inline optax/_src/base_test.py:20]: (Pylint will raise a wrong import order error here)

**Review by mkunesch — COMMENTED** [review]
Hi! The checks still don't pass ... could you sync the latest changes? In theory we fixed this error on Monday. I've also flagged a line which gives a wrong import order error in pylint. 
Once the checks pass I think it's ready to merge! Thanks a lot again!

File: optax/_src/base_test.py (hunk 8 above)
Comment (n2cholas) [reply optax/_src/base_test.py:20]: I'll fix this change, but running `test.sh` locally did not raise this issue for me (tried running `pylint --rcfile=.pylintrc optax/_src/base_test.py` separately as well).

**Review by mkunesch — APPROVED** [review]

File: optax/_src/base_test.py (hunk 8 above)
Comment (mkunesch) [reply optax/_src/base_test.py]: Thanks for letting me know! This would definitely be blocked internally as we try to merge so I'll look into adding an import order check to `test.sh`.

**Comment by mkunesch** [comment]
Thanks a lot again for suggesting and implementing the stateless optimizer - such a great suggestion! We could think about whether to implement a few of the existing gradient transformations in optax using `stateless` ... good targets could be recently added gradient transformations that have good test coverage.

Thanks again!
